python/rikai/spark/functions/vision.py

```diff
@@ -206,6 +215,152 @@ def video_to_images(
     ]
 
 
+video_metadata_schema = StructType(
+    [
+        StructField(
+            "error",
+            StructType(
```

Comment:
so, do we want to advocate this error handling pattern? the one that returns either data or error?

Or , should we just return `data` part , while put `error` into logs? like if anything was wrong, return `null` instead?  It makes it easier to be integrate with other part of the systems, i.e, nested UDFs.

Reply:
is this a commonly used pattern in spark?

Reply:
It's a Scala pattern (try).

Putting the errors in the logs makes it harder to diagnose the error for a big dataset. e.g., i might want to see 1) how many links still point to something 2) how many links point to random files 3) how many links point to audio only tracks, 4) how many files point to video files that are mal-formed/corrupted. 

For ease of use, an expr like "output_col.data.width" will return null for the rows where there are errors and no data so you don't need complicated handling to get a particular attribute if you just want the data.

Reply:
i still dont feel strong with this.  it is like from the relational algebra's aspect, that a table does not only contains data, it also contains "what happened during the time when the data is generated".

in `Apache Beam`, we can do something like

```
data, error = pipeline | Map(func)
```
i dont know what is the good equivalent here.


Can we at least make the schema like:

```
StructType([
        StructField("width", IntegerType(), True),
        StructField("height", IntegerType(), True),
        StructField("num_frames", IntegerType(), True),
        StructField("duration", FloatType(), True),
        StructField("bit_rate", IntegerType(), True),
        StructField("frame_rate", IntegerType(), True),
        StructField("codec", StringType(), True),
        StructField("size", IntegerType(), True),
        StructField("_error", StructType([
                StructField("message", StringType(), True),
                StructField("stderr", StringType(), True),
                StructField("probe", StringType(), True),
            ]
           , True),
        ),
    ]
),
```

In BigQuery, there are a few system added / hidden column named with `_` as prefix.

Reply:
yeah i can do that. I much prefer having clean separation between data and errors, but I recognize that this pattern is only familiar in functional programming and not really for ML/data people as much, esp when they're interacting with DataFrame APIs.